adapters/pangle/pangle.go

```diff
@@ -0,0 +1,201 @@
+package pangle
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type pangleAdapter struct {
+	Endpoint string
+}
+
+type wrappedExtImpBidder struct {
+	*adapters.ExtImpBidder
+	AdType int `json:"adtype,omitempty"`
+}
+
+type pangleBidExt struct {
+	Pangle *bidExt `json:"pangle,omitempty"`
+}
+
+type bidExt struct {
+	AdType *int `json:"adtype,omitempty"`
+}
+
+/* Builder */
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &pangleAdapter{
+		Endpoint: config.Endpoint,
+	}
+
+	return bidder, nil
+}
+
+/* MakeRequests */
+
+func getAdType(imp openrtb.Imp, parsedImpExt *wrappedExtImpBidder) (adType int) {
+	adType = -1
+	// video
+	if imp.Video != nil {
+		if parsedImpExt != nil && parsedImpExt.Prebid != nil && parsedImpExt.Prebid.IsRewardedInventory == 1 {
+			adType = 7
+			return
+		}
+		if imp.Instl == 1 {
+			adType = 8
+			return
+		}
+	}
+	// banner
+	if imp.Banner != nil {
+		if imp.Instl == 1 {
+			adType = 2
+		} else {
+			adType = 1
+		}
+		return
+	}
+	// native
+	if imp.Native != nil && len(imp.Native.Request) > 0 {
+		adType = 5
+		return
+	}
+
+	return
+}
+
+func (pa *pangleAdapter) MakeRequests(request *openrtb.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var requests []*adapters.RequestData
+	var errs []error
+
+	requestCopy := *request
+	for _, imp := range request.Imp {
+		var impExt wrappedExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &impExt); err != nil {
+			errs = append(errs, fmt.Errorf("failed unmarshalling imp ext (err)%s", err.Error()))
+			continue
+		}
+		// detect and fill adtype
+		if adType := getAdType(imp, &impExt); adType == -1 {
+			errs = append(errs, fmt.Errorf("not a supported adtype"))
```

Comment:
You may wish to use a the `BadInput` error type for this case.

Reply:
fixed